### PR TITLE
Fix ovn-kind-cno

### PR DESCRIPTION
**Pull request by trozet**
HA now works by specifying NUM_MASTER_NODES as 3. Updates to CNO
versions so we pull the latest versions of the containers.

Signed-off-by: Tim Rozet <trozet@redhat.com>

### CODE REVIEW

**Comment by trozet** [comment]
@nerdalert @aojea @danielmellado

**Comment by aojea** [comment]
/lgtm
Thanks

File: hack/ovn-kind-cno.sh [hunk 1]
```diff
@@ -90,7 +99,7 @@ if [ "$BUILD_OVN" = true ]; then
   pushd dist/images
   sudo cp -f ../../go-controller/_output/go/bin/* .
   cat << EOF | docker build -t origin-ovn-kubernetes:dev -f - .
-FROM quay.io/openshift/origin-ovn-kubernetes:4.5
+FROM quay.io/openshift/origin-ovn-kubernetes:4.6
```
Comment (abhat) [inline hack/ovn-kind-cno.sh:102]: `quay.io/openshift/origin-ovn-kubernetes:latest` vs. `4.6`?

File: hack/ovn-kind-cno.sh [hunk 2]
```diff
@@ -149,6 +158,12 @@ if [ "$BUILD_OVN" = true ] || [ "$BUILD_CNO" = true ]; then
   popd
 fi
 
+# CNO needs to get scheduled on first node, because thats where KIND will put /etc/kubernetes/kubeconfig
+until kubectl get pod -n openshift-network-operator -o wide | grep "ovn-control-plane ";do
```
Comment (abhat) [inline hack/ovn-kind-cno.sh:162]: exit the script if this operation times out after a few retries?

File: hack/ovn-kind-cno.sh [hunk 3]
```diff
@@ -202,17 +217,20 @@ done
 # wait until resources are created
 sleep 30
 
-if ! kubectl wait -n openshift-ovn-kubernetes --for=condition=ready pods --all --timeout=300s ; then
-  echo "OVN-k8s pods are not running"
+# Wait for pods to come up and be ready. Note metrics pods do not currently work.
```
Comment (abhat) [inline hack/ovn-kind-cno.sh:220]: is there a bz for tracking why metrics pods are not coming up and excluded from the readiness check?

File: hack/ovn-kind-cno.sh (hunk 1 above)
Comment (aojea) [reply hack/ovn-kind-cno.sh:102]: I do not recommend to use latest and better pin the versions, so you always know what you are using and everybody has the same version

File: manifests/0000_70_cluster-network-operator_03_deployment.yaml [hunk 4]
```diff
@@ -16,7 +16,7 @@ spec:
     spec:
       containers:
       - name: network-operator
-        image: quay.io/openshift/origin-cluster-network-operator:4.3
+        image: quay.io/openshift/origin-cluster-network-operator:4.6
```
Comment (abhat) [inline manifests/0000_70_cluster-network-operator_03_deployment.yaml:19]: s/4.6/latest

File: hack/ovn-kind-cno.sh (hunk 1 above)
Comment (Billy99) [reply hack/ovn-kind-cno.sh:102]: I don't know which is right, but @rcarrillocruz push #813 to move them all to latest.

Comment (rcarrillocruz) [reply hack/ovn-kind-cno.sh:102]: Using latest is fine. These scripts are used mainly for development, i.e. master, so it's more than likely that we want to test CNO latest with everything latest.
If we set to 4.6 then it's a moving target, every release we need to change it. Even within the same release, there may be times where a breaking change doesn't work against CNO master.

Comment (aojea) [reply hack/ovn-kind-cno.sh:102]: fair enough,

**Comment by rcarrillocruz** [comment]
As mentioned, I already had up https://github.com/openshift/cluster-network-operator/pull/813/files which tackled the versions of the images in the deployment manifest. So either we land https://github.com/openshift/cluster-network-operator/pull/813/files first and then this just touches the ovn kind script or you change your PR to switch all images to latest and I close my PR.

**Comment by aojea** [comment]
> As mentioned, I already had up https://github.com/openshift/cluster-network-operator/pull/813/files which tackled the versions of the images in the deployment manifest. So either we land https://github.com/openshift/cluster-network-operator/pull/813/files first and then this just touches the ovn kind script or you change your PR to switch all images to latest and I close my PR.

I think ricardos PR is cleaner, since the problem for OVN KIND seems related to the images version and this just adds a new functionality to use HA deployments

File: hack/ovn-kind-cno.sh (hunk 3 above)
Comment (trozet) [reply hack/ovn-kind-cno.sh:220]: well this script isn't really part of the product or anything, I don't know that we need a bz for it. But someone can feel free to fix it in a future patch.

**Comment by trozet** [comment]
> > As mentioned, I already had up https://github.com/openshift/cluster-network-operator/pull/813/files which tackled the versions of the images in the deployment manifest. So either we land https://github.com/openshift/cluster-network-operator/pull/813/files first and then this just touches the ovn kind script or you change your PR to switch all images to latest and I close my PR.
> 
> I think ricardos PR is cleaner, since the problem for OVN KIND seems related to the images version and this just adds a new functionality to use HA deployments

I'm not convinced we want to use latest. Mainly because we use branches in CNO for releases. Therefore after 4.6 branch is pulled and we are on 4.7, if you deploy in 4.6 you will get "latest" which isn't right...it should be 4.6. Yes there is more of an upkeep with updating the versions each release, but on the flip side, when you check out a particular release of the repo, you know what you're getting.

File: hack/ovn-kind-cno.sh (hunk 1 above)
Comment (trozet) [reply hack/ovn-kind-cno.sh:102]: I'd rather keep versions pinned correctly in appropriate branches:
https://github.com/openshift/cluster-network-operator/pull/814#issuecomment-701551960

File: hack/ovn-kind-cno.sh (hunk 2 above)
Comment (trozet) [reply hack/ovn-kind-cno.sh:162]: yeah I can add that to avoid some unforeseen infinite loop

File: hack/ovn-kind-cno.sh (hunk 1 above)
Comment (abhat) [reply hack/ovn-kind-cno.sh:102]: Ok, yeah that makes sense. Let's pin the versions for each branch appropriately. it just becomes a maintenance issue over time. does it make sense to create a version file for each release branch and then pick the version as a tag in each of the clients?

**Comment by trozet** [comment]
/hold

going to modify the versions to detect from git revision

**Comment by aojea** [comment]
@trozet please check this PR https://github.com/openshift/cluster-network-operator/pull/936 too, maybe some of the issues have been identified by @juanluisvaladas  already

**Pull request closed without merging**